src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -185,11 +185,32 @@ namespace detail {
 template <template <typename> class TE, typename ReplacedArgList, size_t I,
           typename TotalContracted, typename T>
 SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
+  // TensorIndex types of pair of indices to contract
+  using lower_tensorindex = ti_contracted_t<I, UpLo::Lo>;
+  using upper_tensorindex = TensorIndex<lower_tensorindex::value + 1, UpLo::Up>;
```

Comment:
why the `+1` here? I don't understand that, sorry. This seems inconsistent with using the same values in `TensorExpression.hpp`

Reply:
I think here it wasn't a bug, but just the convention that was originally in place when I first started working on the contractions (I can't tell which commit this is a comment on from the PR's main page). I believe the convention at this point was to have the first pair of contracted indices (the "replaced args") be `TensorIndex<1000>` and `TensorIndex<1001>`, and if there was another pair, they were `TensorIndex<1002>` and `TensorIndex<1003>`. See the example I gave on the previous PR [here](https://github.com/sxs-collaboration/spectre/pull/2572#discussion_r519142294).

So I believe this `+1` was just a reflection of this convention, not a bug. However, I updated the convention to now be `TensorIndex<1000>` and `TensorIndex<1001>` for the first pair, then `TensorIndex<2000>` and `TensorIndex<2001>` for the next pair as of d6a80bf. This also generalizes for cases where we have > 2 contractions if need be (looking at you, inner products).

Another important note is that I had thought the original intention (when I first started working on the contraction code) was to use this convention not only so that there is a predictable/easy way to find the contracted indices, but also to distinguish the upper and lower indices in the pair - meaning, it seemed intentional that they were given different values, as opposed to both being, for example `TensorIndex<1000>`. I guess now that valence is tracked, now there is less of a need for that, because you can distinguish them through their valences. However, with my updates, one nice thing is that the convention for lower contracted indices is `TensorIndex<1000>`,  `TensorIndex<2000>`, `TensorIndex<3000>`, etc. and upper indices is `TensorIndex<1001>`,  `TensorIndex<2001>`, `TensorIndex<3001>`. So, valence is encoded in the last digit. However, again, I'm not sure what real benefit this has now that we have valence to distinguish between an upper and lower index pair that needs to be contracted (before , upper and lower would have been indistinguishable via type).

Lastly, I also liked the idea of the pairs having different values simply because `evaluate` currently assumes that the list of indices for the RHS and the LHS (separately) are unique. Currently, `LhsSymmAndIndices` in `Evaluate.hpp` does the reordering of the symmetry and indices based on this assumption. Of course, that isn't a reason to keep the convention, but it is a reason why I shamelessly liked keeping the convention while updating this code.

If you think it's better that the upper and lower indices have the same values, let me know! I'll think a bit more on this, too.

Reply:
Okay, thanks for the detailed explanation! I think since the valence is now tracked it would be nice to have them have the same value unless that breaks some code. So feel free to delay that until you have products working :) 

I like that generalized to use 2000, 3000, etc :)